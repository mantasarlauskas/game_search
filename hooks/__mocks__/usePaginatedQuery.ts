import { createRef } from 'react';

const usePaginatedQuery = jest.fn(({ initialData }: any) => ({
    data: initialData,
    isFetching: false,
    paginatorRef: createRef(),
    hasNextPage: true,
    hasPreviousPage: true,
}));

export default usePaginatedQuery;
