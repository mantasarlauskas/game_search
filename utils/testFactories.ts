import { Category } from 'utils/types';

export function getCategories(): Category[] {
    return [
        {
            id: 18893,
            name: 'Feral Interactive',
            slug: 'feral-interactive',
            games_count: 107,
            image_background: 'https://media.rawg.io/media/games/157/15742f2f67eacff546738e1ab5c19d20.jpg',
            games: [
                {
                    slug: 'rise-of-the-tomb-raider',
                    name: 'Rise of the Tomb Raider',
                },
                {
                    slug: 'shadow-of-mordor',
                    name: 'Middle-earth: Shadow of Mordor',
                },
                {
                    slug: 'hitman',
                    name: 'Hitman',
                },
                {
                    slug: 'bioshock-2',
                    name: 'BioShock 2',
                },
                {
                    slug: 'bioshock-remastered',
                    name: 'BioShock Remastered',
                },
                {
                    slug: 'xcom-enemy-unknown',
                    name: 'XCOM: Enemy Unknown',
                },
            ],
        },
        {
            id: 405,
            name: 'Ubisoft',
            slug: 'ubisoft',
            games_count: 297,
            image_background: 'https://media.rawg.io/media/games/5e4/5e4bff02098b2b6f824c68286d5da1a6.jpg',
            games: [
                {
                    slug: 'far-cry-3',
                    name: 'Far Cry 3',
                },
                {
                    slug: 'for-honor',
                    name: 'For Honor',
                },
                {
                    slug: 'watch-dogs',
                    name: 'Watch Dogs',
                },
                {
                    slug: 'far-cry-5',
                    name: 'Far Cry 5',
                },
                {
                    slug: 'assassins-creed-origins',
                    name: 'Assassin\'s Creed Origins',
                },
                {
                    slug: 'rayman-legends',
                    name: 'Rayman Legends',
                },
            ],
        },
    ];
}
