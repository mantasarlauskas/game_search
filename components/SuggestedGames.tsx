import React from 'react';
import { Game } from 'utils/types';
import styles from 'styles/suggested-games.module.scss';
import GameCard from 'components/GameCard';
import PaginatorButton from 'components/PaginatorButton';
import useAppendableResults from 'hooks/useAppendableResults';
import { API_PATH } from 'utils/fetch';

export const suggestedGamesPageSize = 6;

function SuggestedGames({ initialGames, count, name, slug }: SuggestedGamesProps) {
    const {
        results,
        loading,
        paginatorVisible,
        incrementNumber,
    } = useAppendableResults<Game>({
        initialResults: initialGames,
        path: `${API_PATH.GAMES}/${slug}/suggested`,
        count,
        pageSize: suggestedGamesPageSize,
        useIntersectionObserver: false,
    });

    if (!results.length) {
        return null;
    }

    return (
        <div className={styles.root}>
            <div className={styles.title}>
                {`Games like ${name}`}
            </div>
            <div className={styles.cards}>
                {results.map((game) => (
                    <GameCard key={game.slug} game={game} />
                ))}
            </div>
            <div className={styles.button}>
                <PaginatorButton
                    visible={paginatorVisible}
                    loading={loading}
                    onClick={incrementNumber}
                />
            </div>
        </div>
    );
}

interface SuggestedGamesProps {
    initialGames: Game[];
    count: number;
    name: string;
    slug: string;
}

export default SuggestedGames;
