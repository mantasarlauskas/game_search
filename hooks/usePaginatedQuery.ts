import { getFetchUrl, handleServerResponse, QueryParams } from 'utils/fetch';
import { useInfiniteQuery } from 'react-query';
import { DEFAULT_PAGE_SIZE } from 'utils/page';
import { useMemo } from 'react';
import useIntersectionObserver from 'hooks/useIntersectionObserver';

function usePaginatedQuery<T, P extends HTMLElement>({
    initialData,
    path,
    initialNextPage,
    queryParams = {},
    initialPageNumber = 1,
    pageSize = DEFAULT_PAGE_SIZE,
}: UsePaginatedQueryProps<T>) {
    const query = useInfiniteQuery<{
        results: T[],
        previous?: string,
        next?: string,
    }>([path, queryParams], ({
        pageParam = getFetchUrl(path, { page: initialPageNumber, page_size: pageSize, ...queryParams }),
    }) => handleServerResponse(fetch(pageParam)), {
        refetchOnMount: false,
        refetchOnWindowFocus: false,
        getPreviousPageParam: (firstPage, allPages) => allPages[allPages.length - 1]?.previous ?? undefined,
        getNextPageParam: (lastPage) => lastPage?.next ?? undefined,
        initialData: {
            pageParams: [getFetchUrl(path, { page: initialPageNumber, page_size: pageSize, ...queryParams })],
            pages: [{
                results: initialData,
                next: initialNextPage,
            }],
        },
    });

    const paginatorRef = useIntersectionObserver<P>(query.fetchNextPage);
    const data = useMemo(() => query.data?.pages?.reduce((arr, response) => [
        ...arr,
        ...(response?.results || []),
    ], []) || [], [query.data]);

    return {
        ...query,
        data,
        paginatorRef,
    };
}

interface UsePaginatedQueryProps<T> {
    initialData: T[];
    path: string;
    queryParams?: QueryParams;
    initialPageNumber?: number;
    pageSize?: number;
    initialNextPage?: string;
}

export default usePaginatedQuery;
