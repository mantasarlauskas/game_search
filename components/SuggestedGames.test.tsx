import { render, screen } from '@testing-library/react';
import SuggestedGames from 'components/SuggestedGames';
import { getGames } from 'testing/testFactories';
import usePaginatedQuery from 'hooks/usePaginatedQuery';

jest.mock('hooks/usePaginatedQuery');

describe('<SuggestedGames />', () => {
    const props = {
        initialGames: getGames(),
        name: 'Game',
        slug: 'game',
    };

    it('matches snapshot', () => {
        const { container } = render(<SuggestedGames {...props} />);
        expect(container).toMatchSnapshot();
    });

    it('renders content', () => {
        render(<SuggestedGames {...props} />);
        expect(screen.getByText('Portal 2')).toBeInTheDocument();
        expect(screen.getByText('Tomb Raider (2013)')).toBeInTheDocument();
        expect(screen.getByText('Load more')).toBeInTheDocument();
    });

    it('does not render anything', () => {
        (usePaginatedQuery as jest.Mock).mockReturnValue({ data: [] });
        const { container } = render(<SuggestedGames {...props} />);
        expect(container).toBeEmptyDOMElement();
    });
});
