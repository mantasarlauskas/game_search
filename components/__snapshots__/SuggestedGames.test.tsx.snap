// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SuggestedGames /> matches snapshot 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="title"
    >
      Games like Game
    </div>
    <div
      class="cards"
    >
      <div>
        <div
          class="root"
        >
          <div
            class="media"
            role="button"
            style="height: auto;"
            tabindex="0"
          >
            <img
              alt="Portal 2"
              class="image"
              src="https://media.rawg.io/media/crop/600/400/games/328/3283617cb7d75d67257fc58339188742.jpg"
            />
          </div>
          <div
            class="content"
          >
            <div
              class="header"
            >
              <div
                class="title"
              >
                Portal 2
              </div>
              <div
                class="score"
              >
                4.6
              </div>
            </div>
            <div
              class="platforms"
            >
              Platforms: 
              <div>
                <span
                  class="item"
                >
                  PC, 
                </span>
              </div>
              <div>
                <span
                  class="item"
                >
                  Xbox One
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="root"
        >
          <div
            class="media"
            role="button"
            style="height: auto;"
            tabindex="0"
          >
            <img
              alt="Tomb Raider (2013)"
              class="image"
              src="https://media.rawg.io/media/crop/600/400/games/ad2/ad2ffdf80ba993654f31da045bc02456.jpg"
            />
          </div>
          <div
            class="content"
          >
            <div
              class="header"
            >
              <div
                class="title"
              >
                Tomb Raider (2013)
              </div>
              <div
                class="score"
              >
                4.1
              </div>
            </div>
            <div
              class="platforms"
            >
              Platforms: 
              <div>
                <span
                  class="item"
                >
                  PC, 
                </span>
              </div>
              <div>
                <span
                  class="item"
                >
                  Xbox One
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="button"
    >
      <button
        class="root"
        type="button"
      >
        Load more
      </button>
    </div>
  </div>
</div>
`;
